man/1/qmail-tcpsrv.1
.TH qmail-tcpsrv 1
.SH NAME
qmail-tcpsrv \- accept incoming TCP connections
.SH SYNOPSIS
.B qmail-tcpsrv
[
.B \-146jpPhHroOdDqQv
]
[
.B \-c\fIlimit
]
[
.B \-x\fIrules.cdb
]
[
.B \-B\fIbanner
]
[
.B \-g\fIgid
]
[
.B \-u\fIuid
]
[
.B \-b\fIbacklog
]
[
.B \-l\fIlocalname
]
[
.B \-t\fItimeout
]
[
.B \-I\fIinterface
]
.I host
.I port
.I program
[
.I arg ...
]
.SH DESCRIPTION
.B qmail-tcpsrv
is a copy of tcpserver from ucspi-tcp-0.88 (including IPv6). It was integrated in eQmail as a usual requirement.

.B qmail-tcpsrv
waits for connections from TCP clients.
For each connection, it runs
.I program
with the given arguments,
with descriptor 0 reading from the network
and descriptor 1 writing to the network.

The server's address is given by
.I host
and
.IR port .
.I host
can be 0, allowing connections from any host;
or a particular IP address,
allowing connections only to that address;
or a host name, allowing connections to the first IP address
for that host.
.I port
may be a numeric port number
or a port name.
If
.I port
is 0,
.B qmail-tcpsrv
will choose a free port.

.B qmail-tcpsrv
sets up several environment variables,
as described in
.B tcp-environ(5).

.B qmail-tcpsrv
exits when it receives SIGTERM.
.SH "OPTIONS"
.TP
.B \-c\fIlimit
Do not handle more than
.I limit
simultaneous connections.
If there are
.I limit
simultaneous copies of
.I program
running, defer acceptance of a new connection
until one copy finishes.
.I limit
must be a positive integer.
Default: 40.
.TP
.B \-x\fIrules.cdb
Follow the rules compiled into
.I rules.cdb
by
.BR tcprules .
These rules may specify setting environment variables
or rejecting connections from bad sources.

.B qmail-tcpsrv
does not read
.I rules.cdb
into memory;
you can rerun
.B tcprules
to change
.BR qmail-tcpsrv 's
behavior on the fly.
.TP
.B \-B\fIbanner
Write
.I banner
to the network immediately after each connection is made.
.B qmail-tcpsrv
writes
.I banner
before looking up
.BR TCPREMOTEHOST ,
before looking up
.BR TCPREMOTEINFO ,
and before checking
.IR rules.cdb .

This feature can be used to reduce latency in protocols
where the client waits for a greeting from the server.
.TP
.B \-g\fIgid
Switch group ID to
.I gid
after preparing to receive connections.
.I gid
must be a positive integer.
.TP
.B \-u\fIuid
Switch user ID to 
.I uid
after preparing to receive connections.
.I uid
must be a positive integer.
.TP
.B \-1
After preparing to receive connections,
print the local port number to standard output.
.TP
.B \-4
Fall back to IPv4 sockets.  This is necessary for terminally broken
systems like OpenBSD which will not let IPv6 sockets connect to
V4-mapped IPv6 addresses.  Please note that this also applies to DNS
lookups, so you will have to use an DNS resolver with an IPv6 address to
accept IPv6 connections.  Use \fBDNSCACHEIP\fR to set the DNS resolver
IP dynamically.
.TP
.B \-6
Force IPv6 mode in UCSPI environment variables, even for
IPv4 connections.  This will set \fB$PROTO\fR to \fBTCP6\fR and put
IPv4-mapped IPv6 addresses in \fBTCPLOCALIP\fR and \fBTCPREMOTEIP\fR.
.TP
.B \-I\fIinterface
Bind to the network interface
.I interface
("eth0" on Linux, for example).  This is only defined and needed for
IPv6 link-local addresses.
.TP
.B \-b\fIbacklog
Allow up to
.I backlog
simultaneous SYN_RECEIVEDs.
Default: 20.
On some systems,
.I backlog
is silently limited to 5.
See
.BR listen (2)
for more details.
.TP
.B \-o
Leave IP options alone.
If the client is sending packets along an IP source route,
send packets back along the same route.
.TP
.B \-O
(Default.)
Kill IP options.
A client can still use source routing to connect and to send data,
but packets will be sent back along the default route.
.TP
.B \-d
(Default.)
Delay sending data for a fraction of a second whenever the
remote host is responding slowly,
to make better use of the network.
.TP
.B \-D
Never delay sending data;
enable TCP_NODELAY.
This is appropriate for interactive connections.
.TP
.B \-q
Quiet.
Do not print any messages.
.TP
.B \-Q
(Default.)
Print error messages.
.TP
.B \-v
Verbose.
Print all available messages.
.SH "DATA-GATHERING OPTIONS"
.TP
.B \-p
Paranoid.
After looking up the remote host name,
look up the IP addresses for that name,
and make sure one of them matches
.BR TCPREMOTEIP .
If none of them do,
unset
.BR TCPREMOTEHOST .
.TP
.B \-P
(Default.)
Not paranoid.
.TP
.B \-h
(Default.)
Look up the remote host name and set
.BR TCPREMOTEHOST .
.TP
.B \-H
Do not look up the remote host name.
.TP
.B \-l\fIlocalname
Do not look up the local host name;
use
.I localname
for
.BR TCPLOCALHOST .
.TP
.B \-r
(Default: off)
Attempt to obtain
.B TCPREMOTEINFO
from the remote host.
.TP
.B \-t\fItimeout
Give up on the 
.B TCPREMOTEINFO
connection attempt
after
.I timeout
seconds. Default: 26.
.SH "SEE ALSO"
argv0(1),
fixcr(1),
recordio(1),
tcpclient(1),
tcprules(1),
listen(2),
tcp-environ(5)


CHANGES
commit 5d9fb974fa147f29fe7d8fc438ffe9cfb70d803d
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 16:09:15 2017 +0100

    removed subgetopt/sgetopt

commit cabc31d1eae1b9ffc056142afa9f58ad3e1e2ecf
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 13:58:54 2017 +0100

    reworked 'configure' script

commit 6f57aeea75aa12e8fca937de3761d0b2bf0c5d07
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 18:10:04 2017 +0100

    qmail-inject: switched to buffer

commit e3930d23f3df952b6b1276cd7e23722c9dac793f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:59:45 2017 +0100

    qmail-pw2u: switched to 'buffer'

commit bdc2690aecb77ba57aabaf4b4c399ff36cef9789
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:45:29 2017 +0100

    tcp-env: switched to getoptb (buffer)

commit e9d22dd4fff1b4fe8395da4d41862e78036ee53f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 15:36:56 2017 +0100

    preline: switched to 'buffer'

commit ff259662a326a975e56aab07fceae08d9e006298
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 11:25:55 2017 +0100

    added 'socket' lib from ucspi-tcp-088

commit 900d1d7b469ccb793b0b254965523c1f7f3c8a8b
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 08:22:24 2017 +0100

    reworked stralloc lib

commit 71e628955516f4055466902a8e1ca100f56f7ba2
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 13 15:40:54 2017 +0100

    improved build/install system, code clean-up

commit 176589f5c769cc60e9321f17ca3cd294a9f067c0
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 10 11:19:09 2017 +0100

    reworked env lib

commit a0b69812ab0e25089f48b21bc69e5dc5597b0a9f
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:12:38 2017 +0100

    added lib getln_buf

commit 649080aedb1a706827cb1dc935626516cbca9d9e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:11:39 2017 +0100

    install target in Makefile

commit 4a591da4d737131fcc9d10458fed0cbfd9af0c17
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:31:16 2017 +0100

    added fuction 'scan_xlong' to scan lib

commit ed716f2d7587762ed6b1227700e430825f764e03
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:19:04 2017 +0100

    switched to newer sig() lib from ucspi-tcp-0.88

commit d5b1a2055373b39aa59273a74a775a923c120a37
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:06:12 2017 +0100

    added tai, taia and iopauce from ucspi-tcp-0.88

commit 205fbd816da6b4098c748e9b293a1337188acc55
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 10:09:59 2017 +0100

    replaced vfork() by fork()

commit 749e25226bf845a98e87217398ec005ed0ccd983
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 09:12:24 2017 +0100

    updated to tls-patch 20160918

commit 6d1a4dbf3da483e0e9c4beab21538bb75ddedee0
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 16:49:27 2017 +0100

    switch to new(er) cdb lib

commit e4fa52e96d947e02d76d11106a8e614d91d3f28f
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 11:41:55 2017 +0100

    qmail-local: switched to 'getoptb' and 'buffer'

commit 528df12a41c5156c1f05ad83749b0543389a2de5
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 09:38:42 2017 +0100

    added 'getoptb'

commit 83e8b06c4bb1aa8dcf1c54f787bea79b92ee0153
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:18:21 2017 +0100

    added mkrsadhkeys mksrvrcerts

commit 6f739706612ca4d52b6fa4c4fdefe22ed07397c6
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:06:49 2017 +0100

    re-worked fmtqfn.{h,c}

commit 96552e68348125c1f3c1f401780a67b703f8d422
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 11:19:48 2017 +0100

    added 'buffer' lib

commit 19c9c42879ca13d843640dc935f76d3712153c27
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:48:22 2017 +0100

    added qmail-shcfg, qmail-fixq

commit 914ea8122b3591adf48f825eb7f9a6ab9db6690e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:46:48 2017 +0100

    replaced qmail-qstat

commit e5ba1ad6d98f8a9c51c335a220d4c9d0b2a1ad2a
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:20:33 2017 +0100

    replaced outdated install script after rollback

commit 576e6e365f60931ef2276c9b5b1458843e441269
Merge: 446793b e595a09
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:49:36 2016 +0200

    fixed conflicts

commit e595a096efbdd0317e052b57e3c48df8daa51345
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:36:37 2016 +0200

    add files before solve conflicts

commit fc00052ccc96e45be2699bf4fb9fcd224a2b828a
Merge: a76e297 416edce
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:34:46 2016 +0200

    Merge branch 'master' into dev

commit 446793ba603e3fc97a3c460efe2e027809035a23
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:14:51 2016 +0200

    resolve conflicts

commit b651d37a53bde2074cf31020f224658d964e4868
Merge: 360e06b f643417
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:08:28 2016 +0200

    resolve conflicts

commit 360e06b890427969757aa403fc82f4fb947cd58a
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 11:35:14 2016 +0200

    1.09 final

commit a76e297b12ebdf5967e264d36691cc1774c635de
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Apr 8 13:11:17 2016 +0200

    1.10p3

commit f64341737d56aa05fca8695e71f658c6589a6d82
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 17:00:52 2016 +0200

    Update CHANGLOG-1.09

commit 57b16e354477498fc3f6c443bbbfd6b146b9c99c
Merge: 416edce e4c20f9
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:59:32 2016 +0200

    Merge pull request #2 from kp-org/dev

    eQmail-1.09 RC2

commit e4c20f947c4251e23bf619a2158e4d552fbe3165
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:39:57 2016 +0200

    eQmail-1.09 RC2

commit 8a1e087026b1ca66f0bcccff7214833d0bd1aba1
Author: Kai Peter <kp@openqmail.org>
Date:   Sun Mar 27 01:54:09 2016 +0100

    1.10p2

commit cfc7ed9f30bdc01ddc81403c629979ad96562d6f

qmail-tcpsrv.c
/*
 *  Revision 20180412, Kai Peter
 *  - replaced 'strerr' by 'errmsg'
 *  - replaced 'forcev6' and 'fakev4' by 'ipv4socket'
 *  - removed "TCP6*" env vars
 *  - added function 'socket_ipoptionskill'
 *  Revision 20180131, Kai Peter
 *  - changed call of function 'dns_name6()' --> 'dns_name()
 *  Revision 20180119, Kai Peter
 *  - prevent 'misleading-indentation' warning
*/
#include <sys/types.h>
#include <sys/param.h>
#include <netdb.h>
#include <unistd.h>
#include "uint_t.h"
#include "ip.h"
#include "fd.h"
#include "env.h"
#include "prot.h"
#include "open.h"
#include "wait.h"
#include "buffer.h"
#include "errmsg.h"
#include "getoptb.h"
#include "pathexec.h"
#include "qsocket.h"
#include "ndelay.h"
#include "remoteinfo.h"
#include "rules.h"
#include "sig.h"
#include "dns.h"
#include "qstrings.h"

#define WHO "qmail-tcpsrv"

int verbosity = 1;
int flagkillopts = 1;
int flagdelay = 1;
char *banner = "";
int flagremoteinfo = 0;
int flagremotehost = 1;
int flagparanoid = 0;
unsigned long timeout = 26;
uint32 netif = 0;

int ipv4socket;

static stralloc tcpremoteinfo;

uint16 localport;
char localportstr[FMT_ULONG];
char localip[16];
char localipstr[IP6_FMT];
static stralloc localhostsa;
char *localhost = 0;

uint16 remoteport;
char remoteportstr[FMT_ULONG];
char remoteip[16];
char remoteipstr[IP6_FMT];
static stralloc remotehostsa;
char *remotehost = 0;

char strnum[FMT_ULONG];

static stralloc tmp;
static stralloc fqdn;
static stralloc addresses;

char bspace[16];
buffer b;



/* ---------------------------- child */

int flagdeny = 0;
int flagallownorules = 0;
char *fnrules = 0;

void cats(char *s) { if (!stralloc_cats(&tmp,s)) errmem; }
void append(char *ch) { if (!stralloc_append(&tmp,ch)) errmem; }

void safecats(char *s) {
  char ch;
  int i;

  for (i = 0;i < 100;++i) {
    ch = s[i];
    if (!ch) return;
    if (ch < 33) ch = '?';
    if (ch > 126) ch = '?';
    if (ch == '%') ch = '?'; /* logger stupidity */
/*    if (ch == ':') ch = '?'; */
    append(&ch);
  }
  cats("...");
}
void env(const char *s,const char *t) { if (!pathexec_env(s,t)) errmem; }
void drop_rules(void) { errint(111,B("unable to read ",fnrules,": ")); }

/* socket option: ipotionskill */
int socket_ipoptionskill(int s) {
  if (ipv4socket)
    return setsockopt(s,IPPROTO_IP,1,(char *) 0,0); /* 1 == IP_OPTIONS */

  return setsockopt(s,IPPROTO_IPV6,1,(char *) 0,0); /* 1 == IP_OPTIONS */
}

void found(char *data,unsigned int datalen)
{
  unsigned int next0;
  unsigned int split;

  while ((next0 = byte_chr(data,datalen,0)) < datalen) {
    switch(data[0]) {
      case 'D':
    flagdeny = 1;
    break;
      case '+':
    split = str_chr(data + 1,'=');
    if (data[1 + split] == '=') {
      data[1 + split] = 0;
      env(data + 1,data + 1 + split + 1);
    }
    break;
    }
    ++next0;
    data += next0; datalen -= next0;
  }
}

void doit(int t)
{
  int j;
  uint32 scope_id;

  if (ip6_isv4mapped(remoteip))
    remoteipstr[ip4_fmt(remoteipstr,remoteip+12)] = 0;
  else
    remoteipstr[ip6_fmt(remoteipstr,remoteip)] = 0;

  if (verbosity >= 2) {
    strnum[fmt_ulong(strnum,getpid())] = 0;
    log(B("pid ",strnum," from ",remoteipstr,0));
  }

  if (flagkillopts)
    socket_ipoptionskill(t);
  if (!flagdelay)
    socket_tcpnodelay(t);

  if (*banner) {
    buffer_init(&b,write,t,bspace,sizeof bspace);
    if (buffer_putsflush(&b,banner) == -1)
      errint(EHARD,"unable to print banner: ");
  }

  if (socket_local(t,localip,&localport,&scope_id) == -1)
    errint(EHARD,"unable to get local address: ");

  if (ip6_isv4mapped(localip))
    localipstr[ip4_fmt(localipstr,localip+12)] = 0;
  else
    localipstr[ip6_fmt(localipstr,localip)] = 0;
  remoteportstr[fmt_ulong(remoteportstr,remoteport)] = 0;

  if (!localhost)
    if (dns_name(&localhostsa,localip) == 0)
      if (localhostsa.len) {
    if (!stralloc_0(&localhostsa)) errmem;
    localhost = localhostsa.s;
      }
  env("PROTO","TCP");

  env("TCPLOCALIP",localipstr);
  env("TCPLOCALPORT",localportstr);
  env("TCPLOCALHOST",localhost);

  if (flagremotehost)
    if (dns_name(&remotehostsa,remoteip) == 0)
      if (remotehostsa.len) {
    if (flagparanoid)
      if (dns_ip6(&tmp,&remotehostsa) == 0)
        for (j = 0;j + 16 <= tmp.len;j += 16)
          if (byte_equal(remoteip,16,tmp.s + j)) {
        flagparanoid = 0;
        break;
          }
    if (!flagparanoid) {
      if (!stralloc_0(&remotehostsa)) errmem;
      remotehost = remotehostsa.s;
    }
      }
  env("TCPREMOTEIP",remoteipstr);
  env("TCPREMOTEPORT",remoteportstr);
  env("TCPREMOTEHOST",remotehost);

  if (flagremoteinfo) {
    if (remoteinfo6(&tcpremoteinfo,remoteip,remoteport,localip,localport,timeout,netif) == -1)
      flagremoteinfo = 0;
    if (!stralloc_0(&tcpremoteinfo)) errmem;
  }
  env("TCPREMOTEINFO",flagremoteinfo ? tcpremoteinfo.s : 0);

  if (fnrules) {
    int fdrules;
    fdrules = open_read(fnrules);
    if (fdrules == -1) {
      if (errno != error_noent) drop_rules();
      if (!flagallownorules) drop_rules();
    }
    else {
      char* temp;
	  if (ip6_isv4mapped(remoteip))
        temp=remoteipstr+7;
      else
        temp=remoteipstr;
      if (rules(found,fdrules,temp,remotehost,flagremoteinfo ? tcpremoteinfo.s : 0) == -1) drop_rules();
      close(fdrules);
//      log(B("checking tcp rules for ",remotehost,": pass"));
    }
  }

  if (verbosity >= 2) {
    strnum[fmt_ulong(strnum,getpid())] = 0;
    if (!stralloc_copys(&tmp,"qmail-tcpsrv: ")) errmem;
    safecats(flagdeny ? "deny" : "ok");
    cats(" "); safecats(strnum);
    cats(" "); if (localhost) safecats(localhost);
    cats(":"); safecats(localipstr);
    cats(":"); safecats(localportstr);
    cats(" "); if (remotehost) safecats(remotehost);
    cats(":"); safecats(remoteipstr);
//    cats(":<remote info>"); if (flagremoteinfo) safecats(tcpremoteinfo.s);
    cats(":"); if (flagremoteinfo) safecats(tcpremoteinfo.s);
    cats(":"); safecats(remoteportstr);
    cats("\n");
    buffer_putflush(buffer_2,tmp.s,tmp.len);
  }

  if (flagdeny) _exit(100);
}


/* ---------------------------- parent */

void usage(void) {
  errlog(0,ERROR,B("usage: qmail-tcpsrv [ -41UXpPhHrRoOdDqQv ] \
[ -c limit ] [ -x rules.cdb ] [ -B banner ] \
[ -g gid ] [ -u uid ] [ -b backlog ] [ -l localname ] \
[ -t timeout ] [ -I interface ] host port program"));
}

unsigned long limit = 40;
unsigned long numchildren = 0;

int flag1 = 0;
unsigned long backlog = 20;
unsigned long uid = 0;
unsigned long gid = 0;

void printstatus(void) {
  if (verbosity < 2) return;
  log(B("status: ",fmtnum(numchildren),"/",fmtnum(limit)));
}

void sigterm() { _exit(0); }

void sigchld()
{
  int wstat;
  int pid;

  while ((pid = wait_nohang(&wstat)) > 0) {
    if (verbosity >= 2) {
      errlog(0,NOTICE,B("end ",fmtnum(pid)," status ",fmtnum(wstat)));
    }
    if (numchildren) { --numchildren; } printstatus();
  }
}

int main(int argc,char **argv)
{
  char *hostname;
//  char *portname;
  int opt;
  struct servent *se;
  char *x;
  unsigned long u;
  int s;
  int t;

  while ((opt = getopt(argc,argv,"4dDvqQhHrR1UXx:t:u:g:l:b:B:c:I:pPoO")) != opteof)
    switch(opt) {
      case 'b': scan_ulong(optarg,&backlog); break;
      case 'c': scan_ulong(optarg,&limit); break;
      case 'X': flagallownorules = 1; break;
      case 'x': fnrules = optarg; break;
      case 'B': banner = optarg; break;
      case 'd': flagdelay = 1; break;
      case 'D': flagdelay = 0; break;
      case 'v': verbosity = 2; break;
      case 'q': verbosity = 0; break;
      case 'Q': verbosity = 1; break;
      case 'P': flagparanoid = 0; break;
      case 'p': flagparanoid = 1; break;
      case 'O': flagkillopts = 1; break;
      case 'o': flagkillopts = 0; break;
      case 'H': flagremotehost = 0; break;
      case 'h': flagremotehost = 1; break;
//      case 'R': flagremoteinfo = 0; break;
      case 'r': flagremoteinfo = 1; break;
      case 't': scan_ulong(optarg,&timeout); break;
      case 'U': x = env_get("UID"); if (x) scan_ulong(x,&uid);
        x = env_get("GID"); if (x) scan_ulong(x,&gid); break;
      case 'u': scan_ulong(optarg,&uid); break;
      case 'g': scan_ulong(optarg,&gid); break;
      case 'I': netif=socket_getifidx(optarg); break;
      case '1': flag1 = 1; break;
//      case '4': noipv6 = 1; break;
      case '4': ipv4socket = 1; break;
//      case '6': forcev6 = 1; break;
      case 'l': localhost = optarg; break;
      default: usage();
    }
  argc -= optind;
  argv += optind;

  if (!verbosity)
    buffer_2->fd = -1;

  hostname = *argv++;
  if (!hostname) usage();
  if (str_equal(hostname,"")) hostname = "0";

  x = *argv++;
  if (!x) usage();
  if (!x[scan_ulong(x,&u)])
    localport = u;
  else {
    se = getservbyname(x,"tcp");
    if (!se)
      errint(EHARD,B("unable to figure out port number for ",x));
    uint16_unpack_big((char*)&se->s_port,&localport);
  }

  if (!*argv) usage();

  sig_block(sig_child);
  sig_catch(sig_child,sigchld);
  sig_catch(sig_term,sigterm);
  sig_ignore(sig_pipe);

  if (str_equal(hostname,"0")) {
    byte_zero(localip,sizeof localip);
  } else {
    if (!stralloc_copys(&tmp,hostname)) errmem;
    if (dns_ip6_qualify(&addresses,&fqdn,&tmp) == -1)
      errint(EHARD,B("temporarily unable to figure out IP address for ",hostname,": "));
    if (addresses.len < 16)
	  errint(EHARD,B("no IP address for ",hostname));
    byte_copy(localip,16,addresses.s);
    if (ip6_isv4mapped(localip))
      ipv4socket = 1;
  }

  s = socket_tcp();
  if (s == -1)
    errint(EHARD,"unable to create socket: ");
  if (socket_bind_reuse(s,localip,localport,netif) == -1)
    errint(EHARD,"unable to bind: ");

  if (!ipv4socket) ipv4socket = ip6_isv4mapped(localip);

  if (socket_local(s,localip,&localport,&netif) == -1)
    errint(EHARD,"unable to get local address: ");
  if (socket_listen(s,backlog) == -1)
    errint(EHARD,"unable to listen: ");
  ndelay_off(s);

  if (gid) if (prot_gid(gid) == -1)
    errint(EHARD,"unable to set gid: ");
  if (uid) if (prot_uid(uid) == -1)
    errint(EHARD,"unable to set uid: ");


  localportstr[fmt_ulong(localportstr,localport)] = 0;
  if (flag1) {
    buffer_init(&b,write,1,bspace,sizeof bspace);
    buffer_puts(&b,localportstr);
    buffer_puts(&b,"\n");
    buffer_flush(&b);
  }

  close(0);
  close(1);
  printstatus();

  for (;;) {
    while (numchildren >= limit) sig_pause();

    sig_unblock(sig_child);
    t = socket_accept(s,remoteip,&remoteport,&netif);
    sig_block(sig_child);

    if (t == -1) continue;
    ++numchildren; printstatus();

    switch(fork()) {
      case 0:
        close(s);
        doit(t);
        if ((fd_move(0,t) == -1) || (fd_copy(1,0) == -1))
          errint(EHARD,"unable to set up descriptors: ");
        sig_uncatch(sig_child);
        sig_unblock(sig_child);
        sig_uncatch(sig_term);
        sig_uncatch(sig_pipe);
        pathexec(argv);
        errint(EHARD,B("unable to run ",*argv,": "));
      case -1:
        errlog(ESOFT,NOTICE,"unable to fork: ");
        --numchildren; printstatus();
    }
    close(t);
  }
}
